streams/src/test/java/org/apache/kafka/streams/integration/WindowedChangelogRetentionIntegrationTest.java

```diff
@@ -0,0 +1,365 @@
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, windowSize);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveDefaultRetentionWithoutGrace() throws Exception {
+        final Duration windowSize = Duration.ofMillis(500);
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.timeWindowedSerdeFrom(String.class, windowSize.toMillis());
+        groupedStream.windowedBy(TimeWindows.of(Duration.ofMillis(500)))
+                .count(Materialized.as(storeName))
+                .toStream()
+                .to(outputTopic, Produced.with(windowedSerde, Serdes.Long()));
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, DEFAULT_RETENTION);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveDefaultRetentionWithGrace() throws Exception {
+        final Duration windowSize = Duration.ofMillis(500);
+        final Duration grace = Duration.ofMillis(1000);
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.timeWindowedSerdeFrom(String.class, windowSize.toMillis());
+        groupedStream.windowedBy(TimeWindows.of(windowSize).grace(grace))
+                .count(Materialized.as(storeName))
+                .toStream()
+                .to(outputTopic, Produced.with(windowedSerde, Serdes.Long()));
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, DEFAULT_RETENTION);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveRetentionOfWindowSizeAndGrace() throws Exception {
+        final Duration windowSize = Duration.ofDays(1);
+        final Duration grace = Duration.ofHours(12);
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.timeWindowedSerdeFrom(String.class, windowSize.toMillis());
+        groupedStream.windowedBy(TimeWindows.of(windowSize).grace(grace))
+                .count(Materialized.as(storeName))
+                .toStream()
+                .to(outputTopic, Produced.with(windowedSerde, Serdes.Long()));
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, windowSize.plus(grace));
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveUserSpecifiedRetention() throws Exception {
+        final Duration windowSize = Duration.ofDays(1);
+        final Duration grace = Duration.ofHours(12);
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.timeWindowedSerdeFrom(String.class, windowSize.toMillis());
+        groupedStream.windowedBy(TimeWindows.of(windowSize).grace(grace))
+                .count(Materialized.<String, Long, WindowStore<Bytes, byte[]>>as(storeName).withRetention(windowSize.multipliedBy(3)))
```

Comment:
Why "times 3" -- The point is that is must be larger than "max(default, windowSize + grace)" -- maybe hardcode a value instead?

Reply:
"times 3" is indeed not that comprehensible. I just chose a duration that was greater than window size + grace. The point is not that it must be greater than max(default, windowSize + grace). If users specify a retention via `Materialized`, the retention must be larger than window size + grace, otherwhise an exception is thrown. I will set window size + grace + 1 hour.